src/sage/rings/species.py

```diff
@@ -0,0 +1,1468 @@
+            G = libgap.ConjugateGroup(self._group, conj_self)
+            H = libgap.ConjugateGroup(other._group, conj_other)
+            # create dompart
+            dpart = {i + 1: range(x - self._mc[i] + 1, x + 1) for i, x in enumerate(accumulate(self._mc))}
+            # create double coset representatives
+            taus = libgap.DoubleCosetRepsAndSizes(S, G, H)
+            # loop over representatives
+            res = Pn.zero()
+            for tau, _ in taus:
+                F = libgap.Intersection(libgap.ConjugateGroup(H, tau), G)
+                res += Pn(PermutationGroup(gap_group=F, domain=self.domain()), dpart)
+            return res
+
+        def inner_sum(self, base_ring, names, *args):
+            r"""
+            Compute the inner sum of exercise 2.6.16 of BLL book.
+
+            INPUT:
+
+                - ``base_ring``, the base ring of the result
+
+                - ``names``, the names of the result
+
+                - ``args``, the sequence of compositions, each of
+                  which sums to the corresponding cardinality of
+                  ``self``. The number of ``args`` is equal to the
+                  arity of ``self``.
+
+            EXAMPLES::
+
+                sage: P = PolynomialSpecies(ZZ, "X")
+                sage: M = P._indices
+                sage: C4 = M(CyclicPermutationGroup(4))
+                sage: C4.inner_sum(ZZ, "X, Y", [2, 2]) # X^2Y^2 + C2(XY)
+                E_2(XY) + X^2*Y^2
+
+            """
+            # TODO: No checks are performed right now, must be added.
+            # Checks: all args in compositions, sums must match cardinalities.
+
+            # Create group of the composition
+            Pn = PolynomialSpecies(base_ring, names)
+            res = Pn.zero()
+            # conjugate self._group so that [1..k] is sort 1, [k+1,..] is sort 2, so on
+            conj = PermutationGroupElement(list(chain.from_iterable(self._dompart))).inverse()
+            G = libgap.ConjugateGroup(self._group, conj)
+
+            comp = list(chain.from_iterable(args))
+            # Create the double coset representatives.
+            S_down = SymmetricGroup(sum(comp)).young_subgroup(comp)
+            S_up = SymmetricGroup(self._tc).young_subgroup(self._mc)
+            taus = libgap.DoubleCosetRepsAndSizes(S_up, S_down, G)
+            # Sum over double coset representatives.
+            for tau, _ in taus:
+                H = libgap.Intersection(libgap.ConjugateGroup(G, tau), S_down)
+                grp = PermutationGroup(gap_group=H, domain=self.domain())
+                dpart = {i + 1: range(x - comp[i] + 1, x + 1) for i, x in enumerate(accumulate(comp))}
+                res += Pn(grp, dpart)
+            return res
+
+        def __call__(self, *args):
+            r"""
+            Substitute M_1...M_k into self.
+            M_i must all have same arity, same multicardinality,
```

Comment:
why the same multicardinality?  I think, the following should work:
```
sage: M = PolynomialSpecies(QQ, "X, Y")._indices
sage: X = M(SymmetricGroup(1), {1:[1]})
sage: Y = M(SymmetricGroup(1), {2:[1]})
sage: (X*Y)(X, Y^2)
```

Reply:
You're right, it does...

My understanding was that all groups need to have the same multicardinality so that the final result also has the same multicardinality or something like that.

Now I realise that's not the case (though after some thinking, I wonder why I was thinking that anyways...).